test/TestGrains/GenericGrains.cs

```diff
@@ -592,13 +592,26 @@ public Task CancellationTokenCallbackThrow(GrainCancellationToken tc)
             return TaskDone.Done;
         }
 
+        public async Task<bool> CallOtherCancellationTokenCallbackResolve(ILongRunningTaskGrain<T> target)
+        {
+            var tc = new GrainCancellationTokenSource();
+            var grainTask = target.CancellationTokenCallbackResolve(tc.Token);
+            await Task.Delay(300);
+            await tc.Cancel();
+            return await grainTask;
+        }
+
         public Task<bool> CancellationTokenCallbackResolve(GrainCancellationToken tc)
         {
             var tcs = new TaskCompletionSource<bool>();
+            var orleansTs = TaskScheduler.Current;
             tc.CancellationToken.Register(() =>
             {
-                if (Thread.CurrentThread.Name == null ||
-                    !Thread.CurrentThread.Name.Contains("Runtime.Scheduler.WorkerPoolThread"))
+                if (TaskScheduler.Current != orleansTs)
+                {
+                    tcs.SetException(new Exception("Callback executed on wrong scheduler"));
+                }
+                else if (Thread.CurrentThread.Name == null || !Thread.CurrentThread.Name.Contains("Runtime.Scheduler.WorkerPoolThread"))
```

Comment:
Well, that is actually not guaranteed. That is not part of any public API guarantee that we name our threads this way.